clean-modular-code/checks-conditionals/python-function-checks.md

```diff
@@ -26,9 +26,192 @@ Using functions in your data processing pipeline offers several advantages:
+
+```{code-cell} ipython3
+---
+editable: true
+slideshow:
+  slide_type: ''
+tags: [raises-exception]
+---
+def read_file(file_path):
+    try:
+        with open(file_path, 'r') as file:
+            data = file.read()
+        return data
+    except FileNotFoundError:
+        raise FileNotFoundError(f"Oops! I couldn't find the file located at: {file_path}. Please check to see if it exists")
+
+# Raises an error immediately if the file doesn't exist
+file_data = read_file("nonexistent_file.txt")
+```
+
+Notice that the code below doesn't throw an error. This code is allowed to fail quietly.
+This will be problematic for a user as their code will run but they will have no output and will have to carefully debug the code in order to figure out where the problem is.
+
+```{code-cell} ipython3
+from pathlib import Path
+
+def read_file(file_path):
+    """
+    Reads the content of a single file and returns it.
+    Handles errors if the file cannot be read.
+    """
+    try:
+        with open(file_path, 'r') as file:
+            data = file.read()
+            return data
+    except FileNotFoundError:
+        return f"File not found: {file_path}"
+    except Exception as e:
+        return f"An error occurred while reading {file_path}: {e}"
+
+# Example usage
+directory_path = Path("data")
+files = list(directory_path.glob("*.json"))
+for file in files:
+    file_data = read_file(file)
+    print(f"Content of {file.name}:")
+    print(file_data)
+print("All done--I didn't fail but I should have failed quickly")
+```
+
++++ {"editable": true, "slideshow": {"slide_type": ""}}
+
+[Using functions](python-functions) is a great first step in creating a robust
+and maintainable data processing workflow. Functions provide modular units that
+can be tested independently, allowing you to handle various edge cases and
+unexpected scenarios effectively. However, adding checks to your functions
+is the next step towards making your code more robust and maintainable over time.
+
+(pythonic-checks)=
+
+## Make Checks Pythonic
+
+Python has a unique philosophy regarding handling potential errors or
+exceptional cases. This philosophy is often summarized by the acronym EAFP:
+"Easier to Ask for Forgiveness than Permission." When combined with the **fail
+fast** approach, your code can be flexible and resilient to the messy
+realities of data processing.
+
+### EAFP vs. LBYL
+
+There are two main approaches to handling potential errors:
+
+- **LBYL (Look Before You Leap)**: Check for conditions before making calls or
+  accessing data.
+- **EAFP (Easier to Ask for Forgiveness than Permission)**: Assume the operation
+  will succeed and handle any exceptions if they occur.
+
+Pythonic code generally favors the EAFP approach, which allows for **failing
+fast** when an error occurs, providing useful feedback without unnecessary
+checks.
```

Comment:
I'm not sure that this is generally true since it depends on the use case, but YOLO.